x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvSlice.java

```diff
@@ -0,0 +1,294 @@
+            default -> throw EsqlIllegalArgumentException.illegalDataType(field.dataType());
+        };
+    }
+
+    @Override
+    public Object fold() {
+        return EvaluatorMapper.super.fold();
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new MvSlice(source(), newChildren.get(0), newChildren.get(1), newChildren.size() > 2 ? newChildren.get(2) : null);
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, MvSlice::new, field, start, end);
+    }
+
+    @Override
+    public DataType dataType() {
+        return field.dataType();
+    }
+
+    @Override
+    public ScriptTemplate asScript() {
+        throw new UnsupportedOperationException("functions do not support scripting");
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(field, start, end);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null || obj.getClass() != getClass()) {
+            return false;
+        }
+        MvSlice other = (MvSlice) obj;
+        return Objects.equals(other.field, field) && Objects.equals(other.start, start) && Objects.equals(other.end, end);
+    }
+
+    static int calculateOffset(int oldOffset, int fieldValueCount) {
+        return oldOffset < 0 ? oldOffset + fieldValueCount : oldOffset;
+    }
+
+    @Evaluator(extraName = "Int", mvFunction = "true")
+    static void process(IntBlock.Builder builder, int posistion, IntBlock field, int start, int end) {
+        int fieldValueCount = field.getValueCount(posistion);
+        int first = field.getFirstValueIndex(posistion);
+        int startOffset = calculateOffset(start, fieldValueCount);
+        int endOffset = calculateOffset(end, fieldValueCount);
+        if (startOffset > fieldValueCount) builder.appendNull();
+        else {
+            for (int i = startOffset + first; i < Math.min(fieldValueCount, endOffset + 1) + first; i++) {
+                builder.appendInt(field.getInt(i));
+            }
+        }
+    }
+
+    @Evaluator(extraName = "Boolean", mvFunction = "true")
+    static void process(BooleanBlock.Builder builder, int posistion, BooleanBlock field, int start, int end) {
+        int fieldValueCount = field.getValueCount(posistion);
+        int first = field.getFirstValueIndex(posistion);
+        int startOffset = calculateOffset(start, fieldValueCount);
+        int endOffset = calculateOffset(end, fieldValueCount);
+        if (startOffset > fieldValueCount) builder.appendNull();
+        else {
+            for (int i = startOffset + first; i < Math.min(fieldValueCount, endOffset + 1) + first; i++) {
+                builder.appendBoolean(field.getBoolean(i));
+            }
+        }
+    }
+
+    @Evaluator(extraName = "Long", mvFunction = "true")
+    static void process(LongBlock.Builder builder, int posistion, LongBlock field, int start, int end) {
```

Comment:
We can kind of tell from the signature of this method that `start` and `end` have to be single valued and not null. It feels like that's fair to infer.